src/ucs/stats/stats.c

```diff
@@ -149,6 +186,84 @@ static ucs_status_t ucs_stats_node_new(ucs_stats_class_t *cls, ucs_stats_node_t
     return UCS_OK;
 }
 
+static ucs_status_t ucs_stats_filter_node_new(ucs_stats_class_t *cls, ucs_stats_filter_node_t **p_node)
+{
+    ucs_stats_filter_node_t *node;
+
+    node = ucs_malloc(sizeof(ucs_stats_filter_node_t),
+                      "stats filter node");
+    if (node == NULL) {
+        ucs_error("Failed to allocate stats filter node for %s", cls->name);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    *p_node = node;
+    return UCS_OK;
+}
+
+static void ucs_stats_add_wildcard(ucs_stats_node_t *node,
+                                   ucs_stats_filter_node_t* filter_tree)
+{
+    ucs_stats_filter_node_t *temp_filter_node;
+    ucs_stats_filter_node_t *filter_node;
+    ucs_stats_filter_node_t *filter_parent;
+    int found = 0;
+    int filter_index = 0;
+    int i;
+    ucs_status_t status;
+    if (ucs_global_opts.stats_format == UCS_STATS_SUMMARY) {
+        filter_parent = &ucs_stats_context.root_filter_node;
+    } else {
+        filter_parent = node->parent->filter_node;
+    }
+    if (ucs_global_opts.stats_format != UCS_STATS_FULL) {
+        ucs_list_for_each(filter_node, &filter_parent->children, list) {
+            if (!strcmp(filter_node->cls->name, node->cls->name)) {
+                found = 1;
+                break;
+            }
+        }
+    }
+    if (!found) {
+        status = ucs_stats_filter_node_new(node->cls, &filter_node);
+        if (status != UCS_OK) {
+            return;
+        }
+        filter_node->type_list_len = 0;
+        filter_node->ref_count = 0;
+        filter_node->counters_bitmask = 0;
+        ucs_list_head_init(&filter_node->children);
```

Comment:
after extracting common struct, can reuse some of stats_node initialization code here